wind_up/ops_curve_shift.py

```diff
@@ -0,0 +1,239 @@
+        )
+    )
+
+    shift_wind_speed = calculate_curve_shift(
+        curve_shift_input=CurveShiftInput(
+            turbine_name=wtg_name,
+            pre_df=pre_df,
+            post_df=post_df,
+            curve_config=CurveConfig(name=CurveTypes.WIND_SPEED, x_col=pw_col, y_col=scada_ws_col),
+            ops_curve_required_columns=required_cols,
+        )
+    )
+
+    results_dict[f"{CurveTypes.POWER_CURVE.value}_shift"] = shift_power.value
+    results_dict[f"{CurveTypes.RPM.value}_shift"] = shift_rpm.value
+    results_dict[f"{CurveTypes.PITCH.value}_shift"] = shift_pitch.value
+    results_dict[f"{CurveTypes.WIND_SPEED.value}_shift"] = shift_wind_speed.value
+
+    warning_msg = ""
+    for wm in [shift_power.warning_msg, shift_rpm.warning_msg, shift_pitch.warning_msg, shift_wind_speed.warning_msg]:
+        if wm is not None:
+            warning_msg += wm
+
+    if warning_msg:
+        result_manager.warning(warning_msg)
+
+    if plot:
+        compare_ops_curves_pre_post(
+            pre_df=pre_df,
+            post_df=post_df,
+            wtg_name=wtg_name,
+            ws_col=scada_ws_col,
+            pw_col=pw_col,
+            pt_col=pt_col,
+            rpm_col=rpm_col,
+            plot_cfg=plot_cfg,
+            is_toggle_test=(cfg.toggle is not None),
+            sub_dir=sub_dir,
+        )
+
+    return results_dict
+
+
+def _required_cols_are_present(
+    pre_df: pd.DataFrame, post_df: pd.DataFrame, turbine_name: str, required_ops_curve_columns: OpsCurveRequiredColumns
+) -> bool:
+    # check if all required columns are present
+    required_cols = list(required_ops_curve_columns)
+    for req_col in required_cols:
+        if req_col not in pre_df.columns:
+            msg = f"check_for_ops_curve_shift {turbine_name} pre_df missing required column {req_col}"
+            result_manager.warning(msg)
+            return False
+        if req_col not in post_df.columns:
+            msg = f"check_for_ops_curve_shift {turbine_name} post_df missing required column {req_col}"
+            result_manager.warning(msg)
+            return False
+    return True
+
+
+def calculate_curve_shift(curve_shift_input: CurveShiftInput) -> CurveShiftOutput:
+    conf = curve_shift_input.curve_config
+    pre_df = curve_shift_input.pre_df
+    post_df = curve_shift_input.post_df
+    wtg_name = curve_shift_input.turbine_name
+
+    bins = np.arange(0, pre_df[conf.x_col].max() + conf.x_bin_width, conf.x_bin_width) if conf.x_bin_width > 0 else 10  # type: ignore[operator,var-annotated]
+
+    mean_curve = pre_df.groupby(pd.cut(pre_df[conf.x_col], bins=bins, retbins=False), observed=True).agg(
+        x_mean=pd.NamedAgg(column=conf.x_col, aggfunc="mean"),
+        y_mean=pd.NamedAgg(column=conf.y_col, aggfunc="mean"),
+    )
+    post_df["expected_y"] = np.interp(post_df[conf.x_col], mean_curve["x_mean"], mean_curve["y_mean"])
+    mean_df = post_df.mean()
+
+    if conf.name in [CurveTypes.PITCH.value, CurveTypes.PITCH]:
```

Comment:
It seems like it would be better if there were exactly 1 way to identify if the curve shift calculation is for pitch. What about give conf a curvetype attribute which is type CurveTypes so you would have `if conf.curve_type == CurveTypes.PITCH:`

Reply:
Since `CurveTypes` inherits from `str` and `Enum`, then we could simply it to `if conf.name == CurveTypes.PITCH`.